File: MNA_Methods/MATLAB/1-Piece by Piece MNA/MNA_calculator.m
%---------------------------------------------------------------------------------------------------------------------------------------%

% MNA_calculator.m
% 
% Author: Rasul Choupanzadeh
% Date: 05/12/2022
% 
% 
% This code is based on the concepts from [1-2].
% 
% [1] Chung-Wen Ho, A. Ruehli, and P. Brennan. The modified nodal approach to
%     network analysis. IEEE Transactions on Circuits and Systems, 22(6):504–509, 1975
% 
% [2] S. Grivet-Talocia and B. Gustavsen. Passive Macromodeling: Theory and Applications, 1st edition. John Wiley & Sons, 2015.

%---------------------------------------------------------------------------------------------------------------------------------------%


function [Gm,Cm,Gamma,Lv,Bm,M,N]=MNA_calculator(Ra,Rb,L,C,Num_branch)
N=2*Num_branch+1;                      % Number of voltage nodes
M=Num_branch;                          % Number of inductance currents

Ga=1./Ra;
Gb=1./Rb;
Gm=zeros(N,N);
Gm(1,1)=sum(Ga);
nt=1;
for i=2:2:N
    Gm(1,i)=-Ga(nt);
    Gm(i,1)=-Ga(nt);
    Gm(i,i)=Ga(nt);
    Gm(i+1,i+1)=Gb(nt);
    nt=nt+1;
end

Cm=zeros(N,N);
nt=1;
for i=3:2:N
    Cm(i,i)=C(nt);
    nt=nt+1;
end

Lm=diag(L);

Es=zeros(N,M);
nt=1;
for i=2:2:N-1
    Es(i,nt)=1;
    Es(i+1,nt)=-1;
    nt=nt+1;
end
%-------------------------------------Voltage input-----------------------------
Es(1,:)=[];
SEt=Lm\transpose(Es);
Gamma=Es*SEt;

Gm(1,:)=[];
Gm(:,1)=[];
Cm(1,:)=[];
Cm(:,1)=[];
N=N-1;

Bm=zeros(N,1);
nt=1;
for i=1:2:N
    Bm(i,1)=1/Ra(nt);
    nt=nt+1;
end

Lv=zeros(N,M);
nt=1;
for j=1:M
    Lv(nt,j)=1;
    nt=nt+2;
end

end

File: MNA_Methods/Python/1-Piece by Piece MNA/MNA_calculator.py

''' MNA_calculator.py

Author: Rasul Choupanzadeh
Date: 05/12/2022


This code is based on the concepts from [1-2].

[1] Chung-Wen Ho, A. Ruehli, and P. Brennan. The modified nodal approach to
    network analysis. IEEE Transactions on Circuits and Systems, 22(6):504–509, 1975

[2] S. Grivet-Talocia and B. Gustavsen. Passive Macromodeling: Theory and Applications, 1st edition. John Wiley & Sons, 2015.

'''



import numpy as np

def MNA_calculator(Ra, Rb, L, C, Num_branch):
    N = 2*Num_branch+1;
    M = Num_branch;
    Ga = 1/Ra[:];
    Gb = 1/Rb[:];
    Gm = np.zeros(shape=(N,N));
    Gm[0,0] = Ga.sum();
    nt=0;
    for i in range (1,N,2):
        Gm[0,i] = -Ga[nt];
        Gm[i,0] = -Ga[nt];
        Gm[i,i] = Ga[nt];
        Gm[i+1,i+1] = Gb[nt];
        nt = nt+1;
        
    Cm = np.zeros(shape=(N,N));
    nt = 0;
    for i in range (2,N,2):
        Cm[i,i] = C[nt];
        nt = nt+1;
        
    Lm = np.diag(L);
    
    Es = np.zeros(shape=(N,M));
    nt = 0;
    for i in range(1,N-1,2):
        Es[i,nt] = 1;
        Es[i+1,nt] = -1;
        nt = nt+1;
    
    # Voltage input corrections
    Es = np.delete(Es, 0, axis=0);      ## delete the first row of Es
    SEt = np.linalg.inv(Lm) @ np.transpose(Es);
    Gamma = Es @ SEt;
    
    Gm = np.delete(Gm, 0, axis=0);      ## delete the first row of Gm
    Cm = np.delete(Cm, 0, axis=0);      ## delete the first row of Cm
    Gm = np.delete(Gm, 0, axis=1);      ## delete the first column of Gm
    Cm = np.delete(Cm, 0, axis=1);      ## delete the first column of Cm    
    N = N-1;
    
    Bm = np.zeros(shape=(N,1));
    nt = 0;
    for i in range (0,N,2):
        Bm[i,0] = Ga[nt];
        nt = nt+1;
        
    Lv = np.zeros(shape=(N,M));
    nt = 0;
    for i in range(0,M):
        Lv[nt, i] = 1;
        nt = nt+2;
       
    return Gm, Cm, Gamma, Lv, Bm, M, N


